packages/statemanager/src/ethersStateManager.ts

```diff
@@ -0,0 +1,519 @@
+import { blockFromRpc } from '@ethereumjs/block/dist/from-rpc'
+import { RLP } from '@ethereumjs/rlp'
+import { Trie } from '@ethereumjs/trie'
+import {
+  Account,
+  bigIntToHex,
+  bufferToHex,
+  intToHex,
+  isHexPrefixed,
+  toBuffer,
+  unpadBuffer,
+} from '@ethereumjs/util'
+import { JsonRpcProvider, StaticJsonRpcProvider } from '@ethersproject/providers'
+import { debug } from 'debug'
+import { keccak256 } from 'ethereum-cryptography/keccak'
+import { hexToBytes } from 'ethereum-cryptography/utils'
+
+import { BaseStateManager } from '.'
+
+import type { Proof, StateManager } from '.'
+import type { StorageDump } from './interface'
+import type { StorageProof } from './stateManager'
+import type { Common } from '@ethereumjs/common'
+import type { Address, PrefixedHexString } from '@ethereumjs/util'
+
+const log = debug('statemanager')
+
+type blockTagType = bigint | 'latest' | 'earliest'
+export interface EthersStateManagerOpts {
+  provider: string | StaticJsonRpcProvider | JsonRpcProvider
+  blockTag?: blockTagType
+}
+
+export class EthersStateManager extends BaseStateManager implements StateManager {
+  private provider: StaticJsonRpcProvider | JsonRpcProvider
+  private contractCache: Map<string, Buffer>
+  private storageTries: { [key: string]: Trie }
+  // This map tracks which storage slots for each account have been retrieved from the provider.
+  // This ensures that slots retrieved from the provider aren't pulled again and overwrite updates
+  // that occur during the course of running EVM message calls.
+  private externallyRetrievedStorageKeys: Map<string, Map<string, boolean>>
+  private blockTag: string
+  private trie: Trie
+
+  constructor(opts: EthersStateManagerOpts) {
+    super({})
+    // useKeyHashing = true since the web3 api provides proof nodes which are hashed
+    // If there were direct api access to devp2p stack, a normal Trie could have been constructed
+    this.trie = new Trie({ useKeyHashing: true })
+    this.storageTries = {}
+    if (typeof opts.provider === 'string') {
+      this.provider = new StaticJsonRpcProvider(opts.provider)
+    } else if (opts.provider instanceof JsonRpcProvider) {
+      this.provider = opts.provider
+    } else {
+      throw new Error(`valid JsonRpcProvider or url required; got ${opts.provider}`)
+    }
+
+    if (typeof opts.blockTag === 'bigint') {
+      this.blockTag = bigIntToHex(opts.blockTag)
+    } else {
+      this.blockTag = opts.blockTag ?? 'latest'
+    }
```

Comment:
Could we get rid of that duplicate logic and just do this:
```suggestion
this.setBlockTag(opts.blockTag ?? 'latest')
```
might complain about not assigning this.blockTag in the constructor though

Reply:
or use tertiary op: `this.blockTag = (typeof opts.blockTag === 'biging')?bigIntToHex(opts.blockTag): opts.blockTag ?? 'latest'`